### PR TITLE
Added patch to allow to connect via ftp proxy using ssl. Refs #280

### PR DESCRIPTION
**Pull request by fetzerms**
This commit allows to use an encrypted connection via a proxy. Tested with a traditional "open" proxy.

### CODE REVIEW

**Comment by lavv17** [comment]
I think it is possible to activate SSL only for proxies of "open" or "proxy-user@host" types.
To do it correctly, it's necessary to add a new state and linger there until the response for AUTH is received. You cannot send USER until response for AUTH is received. See the lines 1479-1480 in ftpclass.cc for example.

**Comment by fetzerms** [comment]
Thanks for the review!

So, adding a check for the response like:

```
if(conn->auth_sent && !expect->IsEmpty())
	 goto usual_return;
```

And further adjusting the username to use might work? I am currently little unsure with the control flow. I will try to implement/add this when I am back home, later this day. The code currently actually works - but I guess this is, because the response is fast enough and no waiting time is needed (maybe).

**Comment by fetzerms** [comment]
Please review again. I think I got it correctly now :).

File: src/ftpclass.cc [hunk 1]
```diff
@@ -1529,6 +1534,23 @@ int   Ftp::Do()
 		     portname?":":NULL,portname.get(),NULL));
 	    }
 	 }
+#if USE_SSL
+	 if(strcmp(proxy_auth_type,"joined") || strcmp(proxy_auth_type,"joined-acct"))
```
Comment (lavv17) [inline src/ftpclass.cc]: this condition looks incorrect (always true)

File: src/ftpclass.cc [hunk 2]
```diff
@@ -1535,7 +1535,7 @@ int   Ftp::Do()
 	    }
 	 }
 #if USE_SSL
-	 if(strcmp(proxy_auth_type,"joined") || strcmp(proxy_auth_type,"joined-acct"))
+	 if(strcmp(proxy_auth_type,"joined") != 0 && strcmp(proxy_auth_type,"joined-acct") != 0)
```
Comment (lavv17) [inline src/ftpclass.cc]: Ok, it's better now. But AUTH cannot also be used for "user" proxy type. Isn't it better to explicitly list allowed proxy types instead? (they are "open" and "proxy-user@host")

Comment (lavv17) [inline src/ftpclass.cc]: I think it would be nicer to have a "bool use_auth=false" before proxy type checks and set it to true in the "open" and "proxy-user@host" cases, then check the variable and send AUTH if true.
Also it would be nice to avoid the ugly (user && pass) check in two places.

Comment (fetzerms) [inline src/ftpclass.cc]: I added a "auth_allowed" flag to the proxy things. I think:

```
user && pass
proxy_user && proxy_pass
```

Need to be adjusted in another pull request. I think general refactoring might be needed, regarding those often used conditions. Maybe some flag like "has_upw" and "has_proxy_upw".

Comment (lavv17) [inline src/ftpclass.cc]: You are right on the "user&&pass", refactoring should be separate.

**Comment by lavv17** [comment]
Merged. Thank you!
I have also added a change to set the flag in the "proxy-user@host" case.

**Comment by fetzerms** [comment]
Ah, that slipped through ;-). Thank you very much for the merge!
